### PR TITLE
Convert HAL features to bitflags, move legacy out

### PR DESCRIPTION
**Pull request by kvark**
The main reasoning behind using bitflags is not as much an economy of space but the ability to check for intersections and do mathematical set-related operations for Warden. As such, I want to specify which features are affected by a test, so that the framework can skip backends that we know (at run-time) are not capable of running this test. This is way better than current crashing on, say, GL compute.

Requesting early feedback since converting all the backends is considerable amount of work.

### CODE REVIEW

**Comment by kvark** [comment]
@msiglreith this is ready for the review now

File: src/backend/gl/src/device.rs [hunk 1]
```diff
@@ -214,11 +213,28 @@ impl Device {
...
+            (4, 10) => glsl::Version::V4_10,
+            (4, 00) => glsl::Version::V4_00,
+            (3, 30) => glsl::Version::V3_30,
+            (1, 50) => glsl::Version::V1_50,
+            (1, 40) => glsl::Version::V1_40,
+            (1, 30) => glsl::Version::V1_30,
+            (1, 20) => glsl::Version::V1_20,
+            (1, 10) => glsl::Version::V1_10,
+            other if other > (4, 60) => glsl::Version::V4_60,
+            other => panic!("GLSL version is note recognized: {:?}", other),
```
Comment (msiglreith) [inline src/backend/gl/src/device.rs]: 'note'

File: src/backend/dx12/src/lib.rs [hunk 2]
```diff
@@ -780,65 +780,27 @@ impl hal::Instance for Instance {
...
-                    sampler_lod_bias: true,
-                    sampler_border_color: true,
-                },
+                    Features::IMAGE_CUBE_ARRAY |
+                    Features::GEOMETRY_SHADER |
+                    Features::TESSELLATION_SHADER |
+                    //logic_op: false, // Optional on feature level 11_0
+                    Features::MULTI_DRAW_INDIRECT |
+                    Features::FORMAT_BC |
+                    Features::INSTANCE_RATE,
```
Comment (msiglreith) [inline src/backend/dx12/src/lib.rs]: 👍

File: src/backend/gl/src/device.rs [hunk 3]
```diff
@@ -435,9 +438,42 @@ impl d::Device<B> for Device {
...
+        let share = &self.share;
+        descs.iter()
+            .map(|desc| {
+                let program = {
+                    let name = unsafe { gl.CreateProgram() };
+
+                    let shader = self.compile_shader(&desc.shader, pso::Stage::Compute);
+                    unsafe { gl.AttachShader(name, shader) };
+
+                    unsafe { gl.LinkProgram(name) };
```
Comment (msiglreith) [inline src/backend/gl/src/device.rs]: That part should probably be moved into a separate function

File: src/backend/gl/src/info.rs [hunk 4]
```diff
@@ -190,6 +192,41 @@ pub struct Info {
     pub extensions: HashSet<&'static str>,
 }
 
+bitflags! {
+    pub struct LegacyFeatures: u16 {
```
Comment (msiglreith) [inline src/backend/gl/src/info.rs]: Do you want to keep those in GL? If someone wants the target all backends, they need to have access to these flags. We might have dx11 specific flags also.

File: src/warden/src/bin/reftest.rs [hunk 5]
```diff
@@ -185,5 +225,6 @@ fn main() {
         num_failures += harness.run(instance);
     }
     let _ = harness;
+    num_failures += 0;
```
Comment (msiglreith) [inline src/warden/src/bin/reftest.rs]: `+= 0`?

**Comment by kvark** [comment]
Bors r=msiglreith 
I'll follow up with answers

File: src/warden/src/bin/reftest.rs (hunk 5 above)
Comment (kvark) [reply src/warden/src/bin/reftest.rs]: This suppresses rust warning for unused `mut`

File: src/backend/gl/src/info.rs (hunk 4 above)
Comment (kvark) [reply src/backend/gl/src/info.rs]: I want to keep it in GL for now, then we'll think about expressing them in an extension to Vk portability, at which point we'll need to move them back into HAL.

**Comment by kvark** [comment]
miscellaneous fixes added
bors r=msiglreith

**Pull request merged**